File: src/main/java/com/ghostchu/peerbanhelper/util/backgroundtask/BackgroundTask.java
package com.ghostchu.peerbanhelper.util.backgroundtask;

import com.ghostchu.peerbanhelper.text.TranslationComponent;
import lombok.Data;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.time.OffsetDateTime;
@Data
public class BackgroundTask implements AutoCloseable {
    @NotNull
    private TranslationComponent title;
    @Nullable
    private TranslationComponent statusText;
    @NotNull
    private OffsetDateTime startAt;
    @NotNull
    private OffsetDateTime finishedAt;
    @NotNull
    private BackgroundTaskStatus status = BackgroundTaskStatus.QUEUED;
    @NotNull
    private BackgroundTaskProgressBarType barType = BackgroundTaskProgressBarType.INDETERMINATE;
    private boolean disposalImmediatelyAfterComplete = false;
    private long max;
    private long current;

    private final @NotNull BackgroundTaskManager manager;

    public BackgroundTask(@NotNull BackgroundTaskManager manager){
        this.manager = manager;
    }

    @Override
    public void close() throws Exception {
        finishedAt = OffsetDateTime.now();
        if( status == BackgroundTaskStatus.QUEUED || status == BackgroundTaskStatus.RUNNING ){ // 只有这两个状态才变更为 COMPLETED，避免覆盖掉 FAILED 之类的状态
            status = BackgroundTaskStatus.COMPLETED;
        }
    }

    public double getProgress(){
        return max == 0 ? 0.0 : (double) current / (double) max;
    }
}

File: src/main/java/com/ghostchu/peerbanhelper/util/backgroundtask/BackgroundTask.kt
package com.ghostchu.peerbanhelper.util.backgroundtask

import com.ghostchu.peerbanhelper.text.TranslationComponent
import java.time.OffsetDateTime
import java.util.*
import java.util.function.Consumer

abstract class BackgroundTask(var title: TranslationComponent) {
    val id: String = UUID.randomUUID().toString()
    var statusText: TranslationComponent? = null
    val startAt: OffsetDateTime = OffsetDateTime.now()
    var finishedAt: OffsetDateTime? = null
    var status: BackgroundTaskStatus = BackgroundTaskStatus.QUEUED
    var barType: BackgroundTaskProgressBarType = BackgroundTaskProgressBarType.INDETERMINATE
    var isDisposalImmediatelyAfterComplete: Boolean = false
    var max: Long = 0
    var current: Long = 0

    /**
     * Execute the task logic. Call the callback whenever status changes to notify listeners.
     * @param callback callback to invoke when task status/progress changes
     */
    abstract fun start(callback: Consumer<BackgroundTask>)

    /**
     * Mark the task as completed
     */
    fun complete() {
        finishedAt = OffsetDateTime.now()
        if (status == BackgroundTaskStatus.QUEUED || status == BackgroundTaskStatus.RUNNING) {
            status = BackgroundTaskStatus.COMPLETED
        }
    }

    /**
     * Mark the task as failed
     */
    fun fail() {
        finishedAt = OffsetDateTime.now()
        status = BackgroundTaskStatus.FAILED
    }

    val progress: Double
        get() = if (max == 0L) 0.0 else current.toDouble() / max.toDouble()
}
